// @redstone-finance/utils/dist/src/multi-executor/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONFIG = exports.ExecutionMode = void 0;
exports.makeBaseConfig = makeBaseConfig;
const lodash_1 = __importDefault(require("lodash"));
var ExecutionMode;
(function (ExecutionMode) {
    ExecutionMode["RACE"] = "race";
    ExecutionMode["FALLBACK"] = "fallback";
    ExecutionMode["CONSENSUS_MEDIAN"] = "consensus_median";
    ExecutionMode["CONSENSUS_ALL_EQUAL"] = "consensus_all_equal";
    ExecutionMode["AGREEMENT"] = "agreement";
    ExecutionMode["MULTI_AGREEMENT"] = "multi_agreement";
})(ExecutionMode || (exports.ExecutionMode = ExecutionMode = {}));
exports.DEFAULT_CONFIG = {
    agreementQuorumNumber: 2,
    consensusQuorumRatio: 3 / 5,
    defaultMode: ExecutionMode.FALLBACK,
    multiAgreementShouldResolveUnagreedToUndefined: false,
};
function makeBaseConfig(fnDelegateConfig, otherConfig) {
    return lodash_1.default.assign({}, fnDelegateConfig, exports.DEFAULT_CONFIG, otherConfig);
}
