// @redstone-finance/utils/dist/src/common/axios-retry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.axiosGetWithRetries = axiosGetWithRetries;
exports.axiosPostWithRetries = axiosPostWithRetries;
const axios_1 = __importDefault(require("axios"));
const retry_1 = require("./retry");
async function axiosGetWithRetries(url, config, axiosInstance = axios_1.default) {
    return await (0, retry_1.retry)({
        fnName: config.fnName ?? "axios.get",
        fn: async () => await axiosInstance.get(url, {
            timeout: config.timeout,
            headers: config.headers,
            params: config.params,
        }),
        ...config,
    })();
}
async function axiosPostWithRetries(url, data, config, axiosInstance = axios_1.default) {
    return await (0, retry_1.retry)({
        fnName: config.fnName ?? "axios.post",
        fn: async () => await axiosInstance.post(url, data, {
            timeout: config.timeout,
            headers: config.headers,
            params: config.params,
        }),
        ...config,
    })();
}
