// @redstone-finance/utils/dist/src/common/objects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptySet = exports.ensureInitialized = void 0;
exports.mergeObjects = mergeObjects;
exports.getRequiredPropValue = getRequiredPropValue;
exports.isEmpty = isEmpty;
exports.useDefaultIfEmpty = useDefaultIfEmpty;
exports.fetchDefaultIfEmpty = fetchDefaultIfEmpty;
exports.isDefined = isDefined;
exports.isTruthy = isTruthy;
function mergeObjects(objects) {
    return Object.assign({}, ...objects);
}
function getRequiredPropValue(obj, prop) {
    if (obj[prop] === undefined) {
        throw new Error(`Object does not contain required property "${prop}". Obj: ` +
            JSON.stringify(obj));
    }
    return obj[prop];
}
function isEmpty(value) {
    return isDefined(value) && value !== "";
}
function useDefaultIfEmpty(value, defaultValue) {
    return isEmpty(value) ? value : defaultValue;
}
async function fetchDefaultIfEmpty(value, defaultValueFetcher) {
    return isEmpty(value) ? value : await defaultValueFetcher();
}
function isDefined(value) {
    return value !== null && value !== undefined;
}
function isTruthy(value) {
    return !!value;
}
const ensureInitialized = (o, key, initializer) => {
    if (!(key in o)) {
        o[key] = initializer();
    }
    return true;
};
exports.ensureInitialized = ensureInitialized;
const createEmptySet = () => new Set();
exports.createEmptySet = createEmptySet;
