// @redstone-finance/protocol/dist/src/data-package/DataPackage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataPackage = void 0;
const utils_1 = require("ethers/lib/utils");
const redstone_constants_1 = require("../common/redstone-constants");
const Serializable_1 = require("../common/Serializable");
const utils_2 = require("../common/utils");
const data_point_deserializer_1 = require("../data-point/data-point-deserializer");
const SignedDataPackage_1 = require("./SignedDataPackage");
const env = globalThis.process?.env ??
    {};
const allowUnsafeEmptyDataPackages = env.REDSTONE_ALLOW_UNSAFE_EMPTY_DATA_PACKAGES === "true";
const allowUnsafeDataPackagesWithDuplications = env.REDSTONE_ALLOW_UNSAFE_DATA_PACKAGES_WITH_DUPLICATIONS === "true";
class DataPackage extends Serializable_1.Serializable {
    dataPoints;
    timestampMilliseconds;
    dataPackageId;
    constructor(dataPoints, timestampMilliseconds, dataPackageId) {
        super();
        this.dataPoints = dataPoints;
        this.timestampMilliseconds = timestampMilliseconds;
        this.dataPackageId = dataPackageId;
        if (dataPoints.length === 0) {
            (0, utils_2.assert)(allowUnsafeEmptyDataPackages, "Empty data packages are not allowed");
        }
        else {
            const expectedDataPointByteSize = dataPoints[0].getValueByteSize();
            for (const dataPoint of dataPoints) {
                (0, utils_2.assert)(dataPoint.getValueByteSize() === expectedDataPointByteSize, "Values of all data points in a DataPackage must have the same number of bytes");
            }
        }
    }
    getEachDataPointByteSize() {
        if (this.dataPoints.length === 0) {
            (0, utils_2.assert)(allowUnsafeEmptyDataPackages, "Empty data packages are not allowed");
            return 32;
        }
        return this.dataPoints[0].getValueByteSize();
    }
    toBytes() {
        return (0, utils_1.concat)([
            this.serializeDataPoints(),
            this.serializeTimestamp(),
            this.serializeDefaultDataPointByteSize(),
            this.serializeDataPointsCount(),
        ]);
    }
    toObj() {
        return {
            dataPoints: this.dataPoints.map((dataPoint) => dataPoint.toObj()),
            timestampMilliseconds: this.timestampMilliseconds,
            dataPackageId: this.dataPackageId,
        };
    }
    static fromObj(plainObject) {
        const dataPoints = plainObject.dataPoints.map(data_point_deserializer_1.deserializeDataPointFromObj);
        return new DataPackage(dataPoints, plainObject.timestampMilliseconds, plainObject.dataPackageId);
    }
    getSignableHash() {
        const serializedDataPackage = this.toBytes();
        const signableHashHex = (0, utils_1.keccak256)(serializedDataPackage);
        return (0, utils_1.arrayify)(signableHashHex);
    }
    sign(privateKey) {
        const signableHashBytes = this.getSignableHash();
        const signingKey = new utils_1.SigningKey(privateKey);
        const fullSignature = signingKey.signDigest(signableHashBytes);
        return new SignedDataPackage_1.SignedDataPackage(this, fullSignature);
    }
    serializeDataPoints() {
        this.dataPoints.sort((dp1, dp2) => {
            const bytes32dataFeedId1Hexlified = (0, utils_1.hexlify)(dp1.serializeDataFeedId());
            const bytes32dataFeedId2Hexlified = (0, utils_1.hexlify)(dp2.serializeDataFeedId());
            const comparisonResult = bytes32dataFeedId1Hexlified.localeCompare(bytes32dataFeedId2Hexlified);
            (0, utils_2.assert)(comparisonResult !== 0 || allowUnsafeDataPackagesWithDuplications, `Duplicated dataFeedId found: ${dp1.dataFeedId}`);
            return comparisonResult;
        });
        return (0, utils_1.concat)(this.dataPoints.map((dp) => dp.toBytes()));
    }
    serializeTimestamp() {
        return (0, utils_2.convertIntegerNumberToBytes)(this.timestampMilliseconds, redstone_constants_1.TIMESTAMP_BS);
    }
    serializeDataPointsCount() {
        return (0, utils_2.convertIntegerNumberToBytes)(this.dataPoints.length, redstone_constants_1.DATA_POINTS_COUNT_BS);
    }
    serializeDefaultDataPointByteSize() {
        return (0, utils_2.convertIntegerNumberToBytes)(this.getEachDataPointByteSize(), redstone_constants_1.DATA_POINT_VALUE_BYTE_SIZE_BS);
    }
}
exports.DataPackage = DataPackage;
