// @redstone-finance/protocol/dist/src/data-point/DataPoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataPoint = void 0;
const utils_1 = require("ethers/lib/utils");
const Serializable_1 = require("../common/Serializable");
const utils_2 = require("../common/utils");
class DataPoint extends Serializable_1.Serializable {
    dataFeedId;
    value;
    metadata;
    constructor(dataFeedId, value, metadata) {
        super();
        this.dataFeedId = dataFeedId;
        this.value = value;
        this.metadata = metadata;
    }
    serializeDataFeedId() {
        return (0, utils_2.convertStringToBytes32)(this.dataFeedId);
    }
    toObj() {
        return {
            dataFeedId: this.dataFeedId,
            value: utils_1.base64.encode(this.value),
            metadata: this.metadata,
        };
    }
    getValueByteSize() {
        return this.value.length;
    }
    toBytes() {
        const serializedDataFeedId = this.serializeDataFeedId();
        return (0, utils_1.concat)([serializedDataFeedId, this.value]);
    }
}
exports.DataPoint = DataPoint;
