// @redstone-finance/evm-connector/dist/src/WrapperBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrapperBuilder = void 0;
const DataPackagesWrapper_1 = require("./wrappers/DataPackagesWrapper");
const DataServiceWrapper_1 = require("./wrappers/DataServiceWrapper");
const MockWrapper_1 = require("./wrappers/MockWrapper");
const OnDemandRequestWrapper_1 = require("./wrappers/OnDemandRequestWrapper");
const SimpleMockNumericWrapper_1 = require("./wrappers/SimpleMockNumericWrapper");
class WrapperBuilder {
    baseContract;
    constructor(baseContract) {
        this.baseContract = baseContract;
    }
    static wrap(contract) {
        return new WrapperBuilder(contract);
    }
    usingDataService(dataPackagesRequestInput) {
        return new DataServiceWrapper_1.DataServiceWrapper(dataPackagesRequestInput).overwriteEthersContract(this.baseContract);
    }
    usingMockDataPackages(mockDataPackages) {
        return new MockWrapper_1.MockWrapper(mockDataPackages).overwriteEthersContract(this.baseContract);
    }
    usingSimpleNumericMock(simpleNumericMockConfig) {
        return new SimpleMockNumericWrapper_1.SimpleNumericMockWrapper(simpleNumericMockConfig).overwriteEthersContract(this.baseContract);
    }
    usingOnDemandRequest(nodeUrls, scoreType) {
        return new OnDemandRequestWrapper_1.OnDemandRequestWrapper({
            signer: this.baseContract.signer,
            scoreType,
        }, nodeUrls).overwriteEthersContract(this.baseContract);
    }
    usingDataPackages(dataPackages) {
        return new DataPackagesWrapper_1.DataPackagesWrapper(dataPackages).overwriteEthersContract(this.baseContract);
    }
}
exports.WrapperBuilder = WrapperBuilder;
