// @redstone-finance/evm-connector/typechain-types/factories/contracts/libs/SignatureLib__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type { PromiseOrValue } from "../../../common";
import type {
  SignatureLib,
  SignatureLibInterface,
} from "../../../contracts/libs/SignatureLib";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "signedHash",
        type: "bytes32",
      },
    ],
    name: "InvalidSignature",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212209a5d986b517d4da83a92e022cfeeff2b3c3096bdf404077166200d48a4e028e264736f6c63430008110033";

type SignatureLibConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SignatureLibConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SignatureLib__factory extends ContractFactory {
  constructor(...args: SignatureLibConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<SignatureLib> {
    return super.deploy(overrides || {}) as Promise<SignatureLib>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SignatureLib {
    return super.attach(address) as SignatureLib;
  }
  override connect(signer: Signer): SignatureLib__factory {
    return super.connect(signer) as SignatureLib__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SignatureLibInterface {
    return new utils.Interface(_abi) as SignatureLibInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SignatureLib {
    return new Contract(address, _abi, signerOrProvider) as SignatureLib;
  }
}
