// @redstone-finance/evm-connector/typechain-types/contracts/samples/index.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type * as sampleSyntheticTokenSol from "./SampleSyntheticToken.sol";
export type { sampleSyntheticTokenSol };
import type * as benchmarks from "./benchmarks";
export type { benchmarks };
export type { SampleBitmapLib } from "./SampleBitmapLib";
export type { SampleDuplicatedDataFeeds } from "./SampleDuplicatedDataFeeds";
export type { SampleForLocalhostMockTest } from "./SampleForLocalhostMockTest";
export type { SampleKydServiceConsumer } from "./SampleKydServiceConsumer";
export type { SampleNumericArrayLib } from "./SampleNumericArrayLib";
export type { SampleProxyConnector } from "./SampleProxyConnector";
export type { SampleRedstoneConsumerBytesMock } from "./SampleRedstoneConsumerBytesMock";
export type { SampleRedstoneConsumerBytesMockManyDataFeeds } from "./SampleRedstoneConsumerBytesMockManyDataFeeds";
export type { SampleRedstoneConsumerBytesMockStrings } from "./SampleRedstoneConsumerBytesMockStrings";
export type { SampleRedstoneConsumerNumericMock } from "./SampleRedstoneConsumerNumericMock";
export type { SampleRedstoneConsumerNumericMockManyDataFeeds } from "./SampleRedstoneConsumerNumericMockManyDataFeeds";
export type { SampleRedstoneDataServiceConsumerMock } from "./SampleRedstoneDataServiceConsumerMock";
export type { SampleRedstoneDefaultsLib } from "./SampleRedstoneDefaultsLib";
export type { SampleStorageProxy } from "./SampleStorageProxy";
export type { SampleStorageProxyConsumer } from "./SampleStorageProxyConsumer";
export type { SampleWithEvents } from "./SampleWithEvents";
