// @redstone-finance/evm-connector/typechain-types/contracts/samples/SampleKydServiceConsumer.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export interface SampleKydServiceConsumerInterface extends utils.Interface {
  functions: {
    "aggregateValues(uint256[])": FunctionFragment;
    "executeActionPassingKYD()": FunctionFragment;
    "extractTimestampsAndAssertAllAreEqual()": FunctionFragment;
    "getAuthorisedSignerIndex(address)": FunctionFragment;
    "getDataServiceId()": FunctionFragment;
    "getPassedKYDValue()": FunctionFragment;
    "getUniqueSignersThreshold()": FunctionFragment;
    "validateTimestamp(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "aggregateValues"
      | "executeActionPassingKYD"
      | "extractTimestampsAndAssertAllAreEqual"
      | "getAuthorisedSignerIndex"
      | "getDataServiceId"
      | "getPassedKYDValue"
      | "getUniqueSignersThreshold"
      | "validateTimestamp"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "aggregateValues",
    values: [PromiseOrValue<BigNumberish>[]]
  ): string;
  encodeFunctionData(
    functionFragment: "executeActionPassingKYD",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "extractTimestampsAndAssertAllAreEqual",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getAuthorisedSignerIndex",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "getDataServiceId",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getPassedKYDValue",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getUniqueSignersThreshold",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "validateTimestamp",
    values: [PromiseOrValue<BigNumberish>]
  ): string;

  decodeFunctionResult(
    functionFragment: "aggregateValues",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "executeActionPassingKYD",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "extractTimestampsAndAssertAllAreEqual",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getAuthorisedSignerIndex",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getDataServiceId",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getPassedKYDValue",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getUniqueSignersThreshold",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "validateTimestamp",
    data: BytesLike
  ): Result;

  events: {};
}

export interface SampleKydServiceConsumer extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: SampleKydServiceConsumerInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    aggregateValues(
      values: PromiseOrValue<BigNumberish>[],
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    executeActionPassingKYD(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    extractTimestampsAndAssertAllAreEqual(
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { extractedTimestamp: BigNumber }>;

    getAuthorisedSignerIndex(
      signerAddress: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[number]>;

    getDataServiceId(overrides?: CallOverrides): Promise<[string]>;

    getPassedKYDValue(overrides?: CallOverrides): Promise<[boolean]>;

    getUniqueSignersThreshold(overrides?: CallOverrides): Promise<[number]>;

    validateTimestamp(
      receivedTimestampMilliseconds: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[void]>;
  };

  aggregateValues(
    values: PromiseOrValue<BigNumberish>[],
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  executeActionPassingKYD(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  extractTimestampsAndAssertAllAreEqual(
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  getAuthorisedSignerIndex(
    signerAddress: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<number>;

  getDataServiceId(overrides?: CallOverrides): Promise<string>;

  getPassedKYDValue(overrides?: CallOverrides): Promise<boolean>;

  getUniqueSignersThreshold(overrides?: CallOverrides): Promise<number>;

  validateTimestamp(
    receivedTimestampMilliseconds: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<void>;

  callStatic: {
    aggregateValues(
      values: PromiseOrValue<BigNumberish>[],
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    executeActionPassingKYD(overrides?: CallOverrides): Promise<void>;

    extractTimestampsAndAssertAllAreEqual(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getAuthorisedSignerIndex(
      signerAddress: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<number>;

    getDataServiceId(overrides?: CallOverrides): Promise<string>;

    getPassedKYDValue(overrides?: CallOverrides): Promise<boolean>;

    getUniqueSignersThreshold(overrides?: CallOverrides): Promise<number>;

    validateTimestamp(
      receivedTimestampMilliseconds: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    aggregateValues(
      values: PromiseOrValue<BigNumberish>[],
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    executeActionPassingKYD(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    extractTimestampsAndAssertAllAreEqual(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getAuthorisedSignerIndex(
      signerAddress: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getDataServiceId(overrides?: CallOverrides): Promise<BigNumber>;

    getPassedKYDValue(overrides?: CallOverrides): Promise<BigNumber>;

    getUniqueSignersThreshold(overrides?: CallOverrides): Promise<BigNumber>;

    validateTimestamp(
      receivedTimestampMilliseconds: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    aggregateValues(
      values: PromiseOrValue<BigNumberish>[],
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    executeActionPassingKYD(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    extractTimestampsAndAssertAllAreEqual(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getAuthorisedSignerIndex(
      signerAddress: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getDataServiceId(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getPassedKYDValue(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getUniqueSignersThreshold(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    validateTimestamp(
      receivedTimestampMilliseconds: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
